serde/src/java/org/apache/hadoop/hive/serde2/json/HiveJsonReader.java

```diff
@@ -0,0 +1,532 @@
+ * </tr>
+ * </table>
+ */
+public class HiveJsonReader {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(HiveJsonReader.class);
+
+  private final Map<Pair<StructObjectInspector, String>, StructField> discoveredFields =
+      new HashMap<>();
+
+  private final Set<Pair<StructObjectInspector, String>> discoveredUnknownFields =
+      new HashSet<>();
+
+  private final EnumSet<Feature> features = EnumSet.noneOf(Feature.class);
+
+  private final ObjectMapper objectMapper;
+
+  private final TimestampParser tsParser;
+  private BinaryEncoding binaryEncoding;
+  private final ObjectInspector oi;
+
+  /**
+   * Enumeration that defines all on/off features for this reader.
+   */
+  public enum Feature {
+    COL_INDEX_PARSING, PRIMITIVE_TO_WRITABLE, IGNORE_UKNOWN_FIELDS
+  }
+
+  /**
+   * Constructor with default the Hive default timestamp parser.
+   *
+   * @param oi ObjectInspector for all the fields in the JSON object
+   */
+  public HiveJsonReader(ObjectInspector oi) {
+    this(oi, new TimestampParser());
+  }
+
+  /**
+   * Constructor with default the Hive default timestamp parser.
+   *
+   * @param oi ObjectInspector info for all the fields in the JSON object
+   * @param tsParser Custom timestamp parser
+   */
+  public HiveJsonReader(ObjectInspector oi, TimestampParser tsParser) {
+    this.binaryEncoding = BinaryEncoding.BASE64;
+    this.tsParser = tsParser;
+    this.oi = oi;
+    this.objectMapper = new ObjectMapper();
+  }
+
+  /**
+   * Parse text containing a complete JSON object.
+   *
+   * @param text The text to parse
+   * @return A List of Objects, one for each field in the JSON object
+   * @throws IOException Unable to parse the JSON text
+   * @throws SerDeException The SerDe is not configured correctly
+   */
+  public Object parseStruct(final String text)
+      throws IOException, SerDeException {
+    Preconditions.checkNotNull(text);
+    Preconditions.checkState(this.oi != null);
+    final JsonNode rootNode = this.objectMapper.reader().readTree(text);
+    return visitNode(rootNode, this.oi);
+  }
+
+  /**
+   * Parse text containing a complete JSON object.
+   *
+   * @param in The InputStream to read the text from
+   * @return A List of Objects, one for each field in the JSON object
+   * @throws IOException Unable to parse the JSON text
+   * @throws SerDeException The SerDe is not configured correctly
+   */
+  public Object parseStruct(final InputStream in)
+      throws IOException, SerDeException {
+    Preconditions.checkNotNull(in);
+    Preconditions.checkState(this.oi != null);
+    final JsonNode rootNode = this.objectMapper.reader().readTree(in);
+    return visitNode(rootNode, this.oi);
+  }
+
+  /**
+   * Visit a node and parse it based on the provided ObjectInspector.
+   *
+   * @param rootNode The root node to process
+   * @param oi The ObjectInspector to use
+   * @return The value in this node (may be a complex type if nested)
```

Comment:
seems like this can return null, can you add annotation plus explanation what null means ?

Reply:
I can try to clarify,... `null` has not special meaning here, it's just the value of the JsonNode.

Example: `{"name":null}`